File: rosplan_planning_system/src/ActionInterface/RPActionInterface.cpp
#include "rosplan_action_interface/RPActionInterface.h"

/* The implementation of RPMoveBase.h */
namespace KCL_rosplan {

	/* run action interface */
	void RPActionInterface::runActionInterface() {

		ros::NodeHandle nh("~");

		// set action name
		nh.getParam("pddl_action_name", params.name);

		// knowledge base services
		std::string kb = "knowledge_base";
		nh.getParam("knowledge_base", kb);

		// fetch action params
		std::stringstream ss;
		ss << "/" << kb << "/domain/operator_details";
		ros::service::waitForService(ss.str(),ros::Duration(20));
		ros::ServiceClient client = nh.serviceClient<rosplan_knowledge_msgs::GetDomainOperatorDetailsService>(ss.str());
		rosplan_knowledge_msgs::GetDomainOperatorDetailsService srv;
		srv.request.name = params.name;
		if(client.call(srv)) {
			params = srv.response.op.formula;
			op = srv.response.op;
		} else {
			ROS_ERROR("KCL: (RPActionInterface) could not call Knowledge Base for operator details, %s", params.name.c_str());
			return;
		}

		// collect predicates from operator description
		std::vector<std::string> predicateNames;

		// effects
		std::vector<rosplan_knowledge_msgs::DomainFormula>::iterator pit = op.at_start_add_effects.begin();
		for(; pit!=op.at_start_add_effects.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.at_start_del_effects.begin();
		for(; pit!=op.at_start_del_effects.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.at_end_add_effects.begin();
		for(; pit!=op.at_end_add_effects.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.at_end_del_effects.begin();
		for(; pit!=op.at_end_del_effects.end(); pit++)
			predicateNames.push_back(pit->name);

		// simple conditions
		pit = op.at_start_simple_condition.begin();
		for(; pit!=op.at_start_simple_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.over_all_simple_condition.begin();
		for(; pit!=op.over_all_simple_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.at_end_simple_condition.begin();
		for(; pit!=op.at_end_simple_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		// negative conditions
		pit = op.at_start_neg_condition.begin();
		for(; pit!=op.at_start_neg_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.over_all_neg_condition.begin();
		for(; pit!=op.over_all_neg_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		pit = op.at_end_neg_condition.begin();
		for(; pit!=op.at_end_neg_condition.end(); pit++)
			predicateNames.push_back(pit->name);

		// fetch and store predicate details
		ss.str("");
		ss << "/" << kb << "/domain/predicate_details";
		ros::service::waitForService(ss.str(),ros::Duration(20));
		ros::ServiceClient predClient = nh.serviceClient<rosplan_knowledge_msgs::GetDomainPredicateDetailsService>(ss.str());
		std::vector<std::string>::iterator nit = predicateNames.begin();
		for(; nit!=predicateNames.end(); nit++) {
			if (predicates.find(*nit) != predicates.end()) continue;
			if (*nit == "=" || *nit == ">" || *nit == "<" || *nit == ">=" || *nit == "<=") continue;
			rosplan_knowledge_msgs::GetDomainPredicateDetailsService predSrv;
			predSrv.request.name = *nit;
			if(predClient.call(predSrv)) {
				predicates.insert(std::pair<std::string, rosplan_knowledge_msgs::DomainFormula>(*nit, predSrv.response.predicate));
			} else {
				ROS_ERROR("KCL: (RPActionInterface) could not call Knowledge Base for predicate details, %s", params.name.c_str());
				return;
			}
		}

		// create PDDL info publisher
		ss.str("");
		ss << "/" << kb << "/pddl_action_parameters";
		pddl_action_parameters_pub = nh.advertise<rosplan_knowledge_msgs::DomainFormula>(ss.str(), 10, true);

		// create the action feedback publisher
		std::string aft = "default_feedback_topic";
		nh.getParam("action_feedback_topic", aft);
		action_feedback_pub = nh.advertise<rosplan_dispatch_msgs::ActionFeedback>(aft, 10, true);

		// knowledge interface
		ss.str("");
		ss << "/" << kb << "/update_array";
		update_knowledge_client = nh.serviceClient<rosplan_knowledge_msgs::KnowledgeUpdateServiceArray>(ss.str());

		// listen for action dispatch
		std::string adt = "default_dispatch_topic";
		nh.getParam("action_dispatch_topic", adt);

        ros::SubscribeOptions ops;
        ops.template init<rosplan_dispatch_msgs::ActionDispatch>(adt, 1000, boost::bind(&KCL_rosplan::RPActionInterface::dispatchCallback, this, _1));
        ops.transport_hints = ros::TransportHints();
        ops.allow_concurrent_callbacks = true;
		ros::Subscriber ds = nh.subscribe(ops); // nh.subscribe(adt, 1000, &KCL_rosplan::RPActionInterface::dispatchCallback, this);

		// loop
		ros::Rate loopRate(1);
		ros::AsyncSpinner spinner(4);
        spinner.start();

		ROS_INFO("KCL: (%s) Ready to receive", params.name.c_str());

		while(ros::ok()) {
			pddl_action_parameters_pub.publish(params);
			loopRate.sleep();
		}
	}

	/* run action interface */
	void RPActionInterface::dispatchCallback(const rosplan_dispatch_msgs::ActionDispatch::ConstPtr& msg) {

		// check action name
		if(0==msg->name.compare("cancel_action")) {
            action_cancelled = true;
            return;
        }
		if(0!=msg->name.compare(params.name)) return;
		ROS_INFO("KCL: (%s) action received", params.name.c_str());

        action_cancelled = false;

		// check PDDL parameters
		std::vector<bool> found(params.typed_parameters.size(), false);
		std::map<std::string, std::string> boundParameters;
		for(size_t j=0; j<params.typed_parameters.size(); j++) {
			for(size_t i=0; i<msg->parameters.size(); i++) {
				if(params.typed_parameters[j].key == msg->parameters[i].key) {
					boundParameters[msg->parameters[i].key] = msg->parameters[i].value;
					found[j] = true;
					break;
				}
			}
			if(!found[j]) {
				ROS_INFO("KCL: (%s) aborting action dispatch; malformed parameters, missing %s", params.name.c_str(), params.typed_parameters[j].key.c_str());
				return;
			}
		}

		// send feedback (enabled)
		rosplan_dispatch_msgs::ActionFeedback fb;
		fb.action_id = msg->action_id;
		fb.status = "action enabled";
		action_feedback_pub.publish(fb);

		{
			// update knowledge base
			rosplan_knowledge_msgs::KnowledgeUpdateServiceArray updatePredSrv;
			
			// simple START del effects
			for(int i=0; i<op.at_start_del_effects.size(); i++) {
				rosplan_knowledge_msgs::KnowledgeItem item;
				item.knowledge_type = rosplan_knowledge_msgs::KnowledgeItem::FACT;
				item.attribute_name = op.at_start_del_effects[i].name;
				item.values.clear();
				diagnostic_msgs::KeyValue pair;
				for(size_t j=0; j<op.at_start_del_effects[i].typed_parameters.size(); j++) {
					pair.key = predicates[op.at_start_del_effects[i].name].typed_parameters[j].key;
					pair.value = boundParameters[op.at_start_del_effects[i].typed_parameters[j].key];
					item.values.push_back(pair);
				}
				updatePredSrv.request.knowledge.push_back(item);
				updatePredSrv.request.update_type.push_back(rosplan_knowledge_msgs::KnowledgeUpdateService::Request::REMOVE_KNOWLEDGE);
			}

			// simple START add effects
			for(int i=0; i<op.at_start_add_effects.size(); i++) {
				rosplan_knowledge_msgs::KnowledgeItem item;
				item.knowledge_type = rosplan_knowledge_msgs::KnowledgeItem::FACT;
				item.attribute_name = op.at_start_add_effects[i].name;
				item.values.clear();
				diagnostic_msgs::KeyValue pair;
				for(size_t j=0; j<op.at_start_add_effects[i].typed_parameters.size(); j++) {
					pair.key = predicates[op.at_start_add_effects[i].name].typed_parameters[j].key;
					pair.value = boundParameters[op.at_start_add_effects[i].typed_parameters[j].key];
					item.values.push_back(pair);
				}
				updatePredSrv.request.knowledge.push_back(item);
				updatePredSrv.request.update_type.push_back(rosplan_knowledge_msgs::KnowledgeUpdateService::Request::ADD_KNOWLEDGE);
			}

			if(updatePredSrv.request.knowledge.size()>0 && !update_knowledge_client.call(updatePredSrv))
				ROS_INFO("KCL: (%s) failed to update PDDL model in knowledge base", params.name.c_str());
		}

		// call concrete implementation
		action_success = concreteCallback(msg);
        ros::spinOnce();
        if(action_cancelled) {
            action_success = false;
			ROS_INFO("KCL: (%s) an old action that was cancelled is stopping now", params.name.c_str());
            return;
        }

		if(action_success) {

			ROS_INFO("KCL: (%s) action completed successfully", params.name.c_str());

			// update knowledge base
			rosplan_knowledge_msgs::KnowledgeUpdateServiceArray updatePredSrv;

			// simple END del effects
			for(int i=0; i<op.at_end_del_effects.size(); i++) {
				rosplan_knowledge_msgs::KnowledgeItem item;
				item.knowledge_type = rosplan_knowledge_msgs::KnowledgeItem::FACT;
				item.attribute_name = op.at_end_del_effects[i].name;
				item.values.clear();
				diagnostic_msgs::KeyValue pair;
				for(size_t j=0; j<op.at_end_del_effects[i].typed_parameters.size(); j++) {
					pair.key = predicates[op.at_end_del_effects[i].name].typed_parameters[j].key;
					pair.value = boundParameters[op.at_end_del_effects[i].typed_parameters[j].key];
					item.values.push_back(pair);
				}
				updatePredSrv.request.knowledge.push_back(item);
				updatePredSrv.request.update_type.push_back(rosplan_knowledge_msgs::KnowledgeUpdateService::Request::REMOVE_KNOWLEDGE);
			}

			// simple END add effects
			for(int i=0; i<op.at_end_add_effects.size(); i++) {
				rosplan_knowledge_msgs::KnowledgeItem item;
				item.knowledge_type = rosplan_knowledge_msgs::KnowledgeItem::FACT;
				item.attribute_name = op.at_end_add_effects[i].name;
				item.values.clear();
				diagnostic_msgs::KeyValue pair;
				for(size_t j=0; j<op.at_end_add_effects[i].typed_parameters.size(); j++) {
					pair.key = predicates[op.at_end_add_effects[i].name].typed_parameters[j].key;
					pair.value = boundParameters[op.at_end_add_effects[i].typed_parameters[j].key];
					item.values.push_back(pair);
				}
				updatePredSrv.request.knowledge.push_back(item);
				updatePredSrv.request.update_type.push_back(rosplan_knowledge_msgs::KnowledgeUpdateService::Request::ADD_KNOWLEDGE);
			}

			if(updatePredSrv.request.knowledge.size()>0 && !update_knowledge_client.call(updatePredSrv))
				ROS_INFO("KCL: (%s) failed to update PDDL model in knowledge base", params.name.c_str());

			// publish feedback (achieved)
			fb.status = "action achieved";
			action_feedback_pub.publish(fb);

		} else {

			// publish feedback (failed)
			fb.status = "action failed";
			action_feedback_pub.publish(fb);
		}
	}

} // close namespace

File: rosplan_planning_system/src/rosplan_planning_system/ActionInterfacePy/RPActionInterface.py
#!/usr/bin/env python

import rospy, copy
from rosplan_knowledge_msgs.msg import DomainFormula, DomainOperator, KnowledgeItem
from rosplan_dispatch_msgs.msg import ActionDispatch, ActionFeedback
from rosplan_knowledge_msgs.srv import GetDomainOperatorDetailsService, GetDomainPredicateDetailsService
from rosplan_knowledge_msgs.srv import KnowledgeUpdateServiceRequest, KnowledgeUpdateServiceArray
from diagnostic_msgs.msg import KeyValue

class RPActionInterface(object):
    def __init__(self):
        self.params = DomainFormula()
        self.op = DomainOperator()
        self.predicates = {} # str, rosplan_knowledge_msgs.DomainFormula dic
        self.sensed_predicates = {} # str, rosplan_knowledge_msgs.DomainFormula dic
        self.action_feedback_pub = None
        self.update_knowledge_client = None

    def dispatchCallback(self, msg):
        # check action name
        if msg.name == 'cancel_action':
            action_cancelled = True
            return None
        if msg.name != self.params.name:
            return None
        rospy.loginfo('KCL: ({}) action received'.format(self.params.name))

        action_cancelled = False

        # check PDDL parameters
        self.params.typed_parameters
        boundParameters = {}
        found = []
        for i in range(0, len(self.params.typed_parameters)):
            found.append(False)
        for j in range(0, len(self.params.typed_parameters)):
            for i in range(0, len(msg.parameters)):
                if self.params.typed_parameters[j].key == msg.parameters[i].key:
                    boundParameters[msg.parameters[i].key] = msg.parameters[i].value
                    found[j] = True
                    break
            if not found[j]:
                rospy.loginfo('KCL: ({}) aborting action dispatch malformed parameters, missing {}'.format(self.params.name, self.params.typed_parameters[j].key))
                return None

        # send feedback (enabled)
        fb = ActionFeedback()
        fb.action_id = msg.action_id
        fb.status = 'action enabled'
        self.action_feedback_pub.publish(fb)

        # update knowledge base
        updatePredSrv = [[] for i in range(2)] # make list of lists

        # simple START del effects
        for i in range(0, len(self.op.at_start_del_effects)):
            if self.op.at_start_del_effects[i].name in self.sensed_predicates:
                continue # sensed predicate

            item = KnowledgeItem()
            item.knowledge_type = KnowledgeItem.FACT
            item.attribute_name = self.op.at_start_del_effects[i].name
            pair = KeyValue()
            for j in range(0, len(self.op.at_start_del_effects[i].typed_parameters)):
                pair.key = self.predicates[self.op.at_start_del_effects[i].name].typed_parameters[j].key
                pair.value = boundParameters[self.op.at_start_del_effects[i].typed_parameters[j].key]
                item.values.append(copy.deepcopy(pair))
            updatePredSrv[0].append(KnowledgeUpdateServiceRequest.REMOVE_KNOWLEDGE)
            updatePredSrv[1].append(item)

        # simple START add effects
        for i in range(0, len(self.op.at_start_add_effects)):
            if self.op.at_start_add_effects[i].name in self.sensed_predicates:
                continue # sensed predicate

            item = KnowledgeItem()
            item.knowledge_type = KnowledgeItem.FACT
            item.attribute_name = self.op.at_start_add_effects[i].name
            pair = KeyValue()
            for j in range(0, len(self.op.at_start_add_effects[i].typed_parameters)):
                pair.key = self.predicates[self.op.at_start_add_effects[i].name].typed_parameters[j].key
                pair.value = boundParameters[self.op.at_start_add_effects[i].typed_parameters[j].key]
                item.values.append(copy.deepcopy(pair))
            updatePredSrv[0].append(KnowledgeUpdateServiceRequest.ADD_KNOWLEDGE)
            updatePredSrv[1].append(item)

        if (len(updatePredSrv[1]) > 0) and (not self.update_knowledge_client(updatePredSrv[0], updatePredSrv[1])):
            rospy.loginfo('KCL: ({}) failed to update PDDL model in knowledge base'.format(self.params.name))

        # call concrete implementation
        action_success = self.concreteCallback(msg)
        if action_cancelled:
            action_success = False
            rospy.loginfo('KCL: ({}) an old action that was cancelled is stopping now'.format(self.params.name))
            return None

        if action_success:
            rospy.loginfo('KCL: ({}) action completed successfully'.format(self.params.name))

            # update knowledge base
            updatePredSrv = [[] for i in range(2)] # make list of lists

            # simple END del effects
            for i in range(0, len(self.op.at_end_del_effects)):
                if self.op.at_end_del_effects[i].name in self.sensed_predicates:
                    continue # sensed predicate

                item = KnowledgeItem()
                item.knowledge_type = KnowledgeItem.FACT
                item.attribute_name = self.op.at_end_del_effects[i].name
                pair = KeyValue()
                for j in range(0, len(self.op.at_end_del_effects[i].typed_parameters)):
                    pair.key = self.predicates[self.op.at_end_del_effects[i].name].typed_parameters[j].key
                    pair.value = boundParameters[self.op.at_end_del_effects[i].typed_parameters[j].key]
                    item.values.append(copy.deepcopy(pair))
                updatePredSrv[0].append(KnowledgeUpdateServiceRequest.REMOVE_KNOWLEDGE)
                updatePredSrv[1].append(item)

            # simple END add effects
            for i in range(0, len(self.op.at_end_add_effects)):
                if self.op.at_end_add_effects[i].name in self.sensed_predicates:
                    continue # sensed predicate

                item = KnowledgeItem()
                item.knowledge_type = KnowledgeItem.FACT
                item.attribute_name = self.op.at_end_add_effects[i].name
                pair = KeyValue()
                for j in range(0, len(self.op.at_end_add_effects[i].typed_parameters)):
                    pair.key = self.predicates[self.op.at_end_add_effects[i].name].typed_parameters[j].key
                    pair.value = boundParameters[self.op.at_end_add_effects[i].typed_parameters[j].key]
                    item.values.append(copy.deepcopy(pair))
                updatePredSrv[0].append(KnowledgeUpdateServiceRequest.ADD_KNOWLEDGE)
                updatePredSrv[1].append(item)

            if (len(updatePredSrv[1]) > 0) and (not self.update_knowledge_client(updatePredSrv[0], updatePredSrv[1])):
                rospy.loginfo('KCL: ({}) failed to update PDDL model in knowledge base'.format(self.params.name))

            # publish feedback (achieved)
            fb.status = 'action achieved'
            self.action_feedback_pub.publish(fb)
        else:
            # publish feedback (failed)
            fb.status = 'action failed'
            self.action_feedback_pub.publish(fb)

    def runActionInterface(self):
        self.params.name = rospy.get_param('~pddl_action_name', None)

        # knowledge base services
        kb = rospy.get_param('~knowledge_base', 'knowledge_base')

        # fetch action params
        ss = '/' + kb + '/domain/operator_details'
        rospy.wait_for_service(ss, 20)
        client = rospy.ServiceProxy(ss, GetDomainOperatorDetailsService)
        try:
            resp = client(self.params.name)
            self.params = resp.op.formula
            self.op = resp.op
        except:
            rospy.logerr('KCL: (RPActionInterface) could not call Knowledge Base for operator details, {}'.format(self.params.name))
            return None

        # collect predicates from operator description
        predicateNames =[]

        # effects
        for eff in self.op.at_start_add_effects:
            predicateNames.append(eff.name)

        for eff in self.op.at_start_add_effects:
            predicateNames.append(eff.name)

        for eff in self.op.at_start_del_effects:
            predicateNames.append(eff.name)

        for eff in self.op.at_end_add_effects:
            predicateNames.append(eff.name)

        for eff in self.op.at_end_del_effects:
            predicateNames.append(eff.name)

        # simple conditions
        for cond in self.op.at_start_simple_condition:
            predicateNames.append(cond.name)

        for cond in self.op.over_all_simple_condition:
            predicateNames.append(cond.name)

        for cond in self.op.at_end_simple_condition:
            predicateNames.append(cond.name)

        # negative conditions
        for ncond in self.op.at_start_neg_condition:
            predicateNames.append(ncond.name)

        for ncond in self.op.over_all_neg_condition:
            predicateNames.append(ncond.name)

        for ncond in self.op.at_end_neg_condition:
            predicateNames.append(ncond.name)

        # fetch and store predicate details
        ss = '/' + kb + '/domain/predicate_details'
        rospy.wait_for_service(ss, 20)
        predClient = rospy.ServiceProxy(ss, GetDomainPredicateDetailsService)
        for predicateName in predicateNames:
            if predicateName in self.predicates:
                continue
            if predicateName == '=' or predicateName == '>' or predicateName == '<' or predicateName == '>=' or predicateName == '<=':
                continue
            try:
                resp = predClient(predicateName)
                if resp.is_sensed:
                    self.sensed_predicates[predicateName] = resp.predicate
                else:
                    self.predicates[predicateName] = resp.predicate
            except:
                rospy.logerr('KCL: (RPActionInterface) could not call Knowledge Base for predicate details, {}'.format(self.params.name))
                return None

        # create PDDL info publisher
        ss = '/' + kb + '/pddl_action_parameters'
        pddl_action_parameters_pub = rospy.Publisher(ss, DomainFormula, queue_size=10)

        # create the action feedback publisher
        aft = rospy.get_param('~action_feedback_topic', 'default_feedback_topic')
        self.action_feedback_pub = rospy.Publisher(aft, ActionFeedback, queue_size=10)

        # knowledge interface
        ss = '/' + kb + '/update_array'
        rospy.wait_for_service(ss, 20)
        self.update_knowledge_client = rospy.ServiceProxy(ss, KnowledgeUpdateServiceArray)

        # listen for action dispatch
        adt = rospy.get_param('~action_dispatch_topic', 'default_dispatch_topic')
        rospy.Subscriber(adt, ActionDispatch, self.dispatchCallback, queue_size=1)
        rospy.loginfo('KCL: ({}) Ready to receive'.format(self.params.name))

        # loop
        loopRate = rospy.Rate(1.0)
        while not rospy.is_shutdown():
            pddl_action_parameters_pub.publish(self.params)
            loopRate.sleep()
